### PR TITLE
IFU TE 2.8.0.dev0 commit 7f77127 from 2025-09-18

### PR DESCRIPTION
**Pull request by ipanfilo**
# Description

IFU TE 2.8.0.dev0 commit 7f77127 from 2025-09-18

Fixes [#14813](https://github.com/ROCm/frameworks-internal/issues/14813)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/pytorch/attention/test_attention.py [hunk 1]
```diff
@@ -375,10 +375,6 @@ def test_dpa_checkpoint(dtype, model_configs, model):
     "mla_3_2": ModelConfig(8, 1, 16, 192, max_seqlen_kv=2048, head_dim_v=128),  # inference
     "mla_3_3": ModelConfig(8, 1, 16, 160, max_seqlen_kv=2048, head_dim_v=128),  # inference
     "mla_3_4": ModelConfig(8, 1, 16, 160, max_seqlen_kv=2048, head_dim_v=160),  # inference
-    #"mla_4_0": ModelConfig(#PIV TODO: do cross 0 and cross 1 cover it
```
Comment (wenchenvincent) [inline tests/pytorch/attention/test_attention.py]: Are these two configs not supported on ROCm?

File: tests/pytorch/attention/test_kv_cache.py [hunk 2]
```diff
@@ -386,6 +386,12 @@ def get_tols(config, module, backend, dtype):
                     torch.half: (1e-2, 1e-2),
                     torch.bfloat16: (8e-2, 7e-2),
                 }
+            # With FA on ROCm it may not fit default tolerance
```
Comment (wenchenvincent) [inline tests/pytorch/attention/test_kv_cache.py:389]: Is this update due to some changes from upstream?

File: tests/pytorch/test_numerics.py [hunk 3]
```diff
@@ -2972,6 +2984,9 @@ def test_fp8gemm_with_unfused_quantization(N, datatype, input_quantizer, out_qua
         pytest.skip(reason_for_no_fp8)
     if is_mxfp8_needed and not mxfp8_available:
         pytest.skip(reason_for_no_mxfp8)
+    if IS_HIP_EXTENSION and get_device_compute_capability() == (9, 5):
```
Comment (wenchenvincent) [inline tests/pytorch/test_numerics.py:2957]: Is this config in the filed ticket?

File: transformer_engine/pytorch/tensor/mxfp8_tensor.py [hunk 4]
```diff
@@ -110,7 +112,8 @@ def make_empty(
 
         # Allocate FP8 data
         data = torch.empty(shape, dtype=torch.uint8, device=device)
-        scale_inv = torch.empty(
+        # ROCm TE does not implement fuse padding zeros so use zero tensor here
+        scale_inv = torch.zeros(
```
Comment (wenchenvincent) [inline transformer_engine/pytorch/tensor/mxfp8_tensor.py:116]: Why the `scale_inv` need to be initialized to 0? Did the tests fail without this?

Comment (ipanfilo) [reply transformer_engine/pytorch/tensor/mxfp8_tensor.py:116]: It has padding that won't be properly filled so yes, test fails w/o it. Using torch.zeros() was an original implementation and it switched to torch.empty() when fuse padding zeros was implemented in upstream

File: tests/pytorch/test_numerics.py (hunk 3 above)
Comment (ipanfilo) [reply tests/pytorch/test_numerics.py:2957]: Not yet

File: tests/pytorch/attention/test_kv_cache.py (hunk 2 above)
Comment (ipanfilo) [reply tests/pytorch/attention/test_kv_cache.py:389]: This is what I do not know. This test seems passing on TE 2.6 with the same FA so it could be some change in TE or it could be just luck - failure appears in one element from ~4k

File: tests/pytorch/attention/test_attention.py (hunk 1 above)
Comment (ipanfilo) [reply tests/pytorch/attention/test_attention.py]: Those are 192x128 configs added specifically on ROCm but later upstream added mla_1_2 and mla_2_2 that are 192x128

File: transformer_engine/pytorch/tensor/mxfp8_tensor.py (hunk 4 above)
Comment (wenchenvincent) [reply transformer_engine/pytorch/tensor/mxfp8_tensor.py:116]: What's the padding for?

File: tests/pytorch/test_numerics.py (hunk 3 above)
Comment (wenchenvincent) [reply tests/pytorch/test_numerics.py:2957]: Let's gather those and file a ticket.

Comment (ipanfilo) [reply tests/pytorch/test_numerics.py:2957]: Configs are gathered is it just no ticket cause there are some more MXFP8 ones not related to IFU

File: transformer_engine/pytorch/tensor/mxfp8_tensor.py (hunk 4 above)
Comment (ipanfilo) [reply transformer_engine/pytorch/tensor/mxfp8_tensor.py:116]: Padding is created with some sizes alignment that results in having padding for configs with matrices sizes not multiply of 32x like 32768x160.

Comment (wenchenvincent) [reply transformer_engine/pytorch/tensor/mxfp8_tensor.py:116]: It is a bit strange to set scale to 0 (meaning 2^-127) for padding as the scale is shared by 32 elements. The padding should be setting the data elements to 0, shouldn't it?

File: tests/cpp/operator/test_cast_mxfp8.cu [hunk 5]
```diff
@@ -311,17 +311,26 @@ void performTest_x1(const ProcessingMethod processing_method,
                                            : output_c.columnwise_cpu_scale_inv_ptr<fp8e8m0>();
 
     const size_t scale_diff_abs_tolerance = 0;
-    const double abs_tolerable_mismatches_limit = 0.0;
-    const double rel_tolerable_mismatches_limit = 0.0;
+    const double abs_tolerable_mismatches_limit = 1.0;
```
Comment (wangye805) [inline tests/cpp/operator/test_cast_mxfp8.cu:315]: Let's guard our rocm specific changes by ifdef?

File: tests/cpp/operator/test_cast_mxfp8.cu [hunk 6]
```diff
@@ -482,27 +491,43 @@ void performTest_x2(const ProcessingMethod processing_method,
                                        scales_stride_colwise);
 
     const size_t scale_diff_abs_tolerance = 0;
-    const double abs_tolerable_mismatches_limit = 0.0;
-    const double rel_tolerable_mismatches_limit = 0.0;
+    const double abs_tolerable_mismatches_limit = 1.0;
```
Comment (wangye805) [inline tests/cpp/operator/test_cast_mxfp8.cu:500]: Same here, need to guard our rocm specific behavior by ifdef

File: tests/jax/test_distributed_layernorm_mlp.py [hunk 7]
```diff
@@ -231,7 +231,7 @@ def _test_layernorm_mlp_grad(
 
                 multi_fwd, multi_grads = multi_jitter(*multi_inputs, *static_inputs, True)
         
-        # TODO: skip cases with single fwd as nan/inf//PIV TODO: is it AMD path?
+        # TODO: skip cases with single fwd as nan/inf
```
Comment (wangye805) [inline tests/jax/test_distributed_layernorm_mlp.py:238]: This is from our code change. 

People found it out in NV upstream as well with an open ticket https://github.com/NVIDIA/TransformerEngine/issues/2415

File: tests/jax/test_distributed_softmax.py [hunk 8]
```diff
@@ -135,6 +135,7 @@ def impl_test_softmax(
                             f"{str(w)}"
                         )
 
+    @pytest.mark.skipif(version.parse(jax.__version__) < version.parse("0.5.0"), reason="shardy sharding requires JAX 0.5.0")
```
Comment (wangye805) [inline tests/jax/test_distributed_softmax.py]: Do we still support jax <= 0.5.0? I saw you removed similar lines in line 179

File: tests/pytorch/attention/test_attention.py [hunk 9]
```diff
@@ -376,12 +375,10 @@ def test_dpa_checkpoint(dtype, model_configs, model):
     "mla_3_2": ModelConfig(8, 1, 16, 192, max_seqlen_kv=2048, head_dim_v=128),  # inference
     "mla_3_3": ModelConfig(8, 1, 16, 160, max_seqlen_kv=2048, head_dim_v=128),  # inference
     "mla_3_4": ModelConfig(8, 1, 16, 160, max_seqlen_kv=2048, head_dim_v=160),  # inference
-    "mla_4_0": ModelConfig( #PIV TODO
-        10, 16, 16, 192, 4096, 4096, 0.0, "causal", "no_bias", head_dim_v=128
-    ),
-    "mla_4_1": ModelConfig(
-        10, 16, 16, 192, 4096, 4096, 0.0, "no_mask", "no_bias", head_dim_v=128
-    ),
+    #"mla_4_0": ModelConfig(#PIV TODO: do cross 0 and cross 1 cover it
```
Comment (wangye805) [inline tests/pytorch/attention/test_attention.py]: Yes. Those two can be covered cross 0 and cross 1 completely.

File: tests/pytorch/attention/test_kv_cache.py [hunk 10]
```diff
@@ -1,3 +1,5 @@
+# This file was modified for portability to AMDGPU
+# Copyright (c) 2026, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (wangye805) [inline tests/pytorch/attention/test_kv_cache.py:2]: I recall we have some pytest failures for test_kv_cache before. Now we can run through it?

File: tests/pytorch/test_numerics.py [hunk 11]
```diff
@@ -163,6 +163,8 @@ def is_fused_attn_available(
         is_training=is_training,
         deterministic=deterministic,
     )
+    if IS_HIP_EXTENSION:
```
Comment (wangye805) [inline tests/pytorch/test_numerics.py:168]: get_available_attention_backends will overwrite our envs for flash-attn and fused-attn. Do you want to have similar change as my IFU v2.6 release branch which moves EnvVarCleaner to util.py, import and use it here?

File: transformer_engine/common/swizzle/swizzle.cu
Comment (wangye805) [inline transformer_engine/common/swizzle/swizzle.cu]: 2026 instead of 2025

File: transformer_engine/common/util/rocm_cast_gated_kernels.cuh [hunk 12]
```diff
@@ -185,20 +187,31 @@ __global__ void __launch_bounds__(THREADS_PER_CHUNK)
           act_x = ActOP(x, {});
           dact_x = DActOP(x, {});
         }
-        after_dact_reg[stage] = dact_x * grad_elt * gate_elt;
```
Comment (wangye805) [inline transformer_engine/common/util/rocm_cast_gated_kernels.cuh:188]: What's the difference before and after your code change of this section?
```
-        after_dact_reg[stage] = dact_x * grad_elt * gate_elt;
-        after_dgate_reg[stage] = act_x * grad_elt;
+        after_act_elt = dact_x * grad_elt * gate_elt;
+        after_gate_elt = act_x * grad_elt;
+        after_dact_reg[stage] = after_act_elt;
+        after_dgate_reg[stage] = after_gate_elt;
```
It looks like all data are defined to be float32

File: transformer_engine/common/util/rocm_cast_gated_kernels.cuh [hunk 13]
```diff
@@ -185,20 +187,31 @@ __global__ void __launch_bounds__(THREADS_PER_CHUNK)
...
-        after_dgate_reg[stage] = act_x * grad_elt;
+        after_act_elt = dact_x * grad_elt * gate_elt;
+        after_gate_elt = act_x * grad_elt;
+        after_dact_reg[stage] = after_act_elt;
+        after_dgate_reg[stage] = after_gate_elt;
       } else {
-        after_dact_reg[stage] = ActOP(act_elt, {}) * gate_elt;
+        after_act_elt = ActOP(act_elt, {}) * gate_elt;
```
Comment (wangye805) [inline transformer_engine/common/util/rocm_cast_gated_kernels.cuh]: Same here, not seeing any logical difference before and after

File: transformer_engine/common/util/rocm_cast_gated_kernels.cuh [hunk 14]
```diff
@@ -185,20 +187,31 @@ __global__ void __launch_bounds__(THREADS_PER_CHUNK)
...
+        if constexpr (IS_DGATED) {
+          after_gate_elt = static_cast<float>(static_cast<IType>(after_gate_elt));
+        }
       }
 
       if constexpr (USE_ROWWISE_SCALING) {
         if constexpr (IS_DGATED) {
           // dgate
-          float amax = fabsf(after_dgate_reg[stage]);
+          float amax = fabsf(after_gate_elt);
```
Comment (wangye805) [inline transformer_engine/common/util/rocm_cast_gated_kernels.cuh]: Is after_gate_elt same as after_dgate_reg[stage]?

File: transformer_engine/common/util/rocm_cast_gated_kernels.cuh [hunk 15]
```diff
@@ -214,11 +227,11 @@ __global__ void __launch_bounds__(THREADS_PER_CHUNK)
             scales_rowwise[scale_idx] = biased_exponent_X;
           }
         }
-        float amax = fabsf(after_dact_reg[stage]);
+        float amax = fabsf(after_act_elt);
```
Comment (wangye805) [inline transformer_engine/common/util/rocm_cast_gated_kernels.cuh]: Emm, you can do the downcast to after_dact_reg[stage] directly without adding after_act_elt. This can minimize the code change and also reduce the possibility of register spill

**Review by wangye805 — CHANGES_REQUESTED** [review]

File: transformer_engine/pytorch/tensor/mxfp8_tensor.py (hunk 4 above)
Comment (ipanfilo) [reply transformer_engine/pytorch/tensor/mxfp8_tensor.py:116]: Padding is not filled with anything during computation. So if triton and hip tensors are created with torch.empty() they end up with different values. Potentially any value can be used for initialization, zero is used because it was used originally.

File: tests/pytorch/test_numerics.py (hunk 3 above)
Comment (ipanfilo) [reply tests/pytorch/test_numerics.py:2957]: SWDEV-575729

File: tests/pytorch/attention/test_kv_cache.py (hunk 10 above)
Comment (ipanfilo) [reply tests/pytorch/attention/test_kv_cache.py:2]: Failed ones are filtered out

File: transformer_engine/common/util/rocm_cast_gated_kernels.cuh (hunk 12 above)
Comment (ipanfilo) [reply transformer_engine/common/util/rocm_cast_gated_kernels.cuh:188]: The changes are to get after_act_elt and after_gate_elt that are used below

File: transformer_engine/common/util/rocm_cast_gated_kernels.cuh (hunk 13 above)
Comment (ipanfilo) [reply transformer_engine/common/util/rocm_cast_gated_kernels.cuh]: The same reason as above

File: transformer_engine/common/util/rocm_cast_gated_kernels.cuh (hunk 14 above)
Comment (ipanfilo) [reply transformer_engine/common/util/rocm_cast_gated_kernels.cuh]: No, it is not the same. Numerical truncation is performed above

File: transformer_engine/common/util/rocm_cast_gated_kernels.cuh (hunk 15 above)
Comment (ipanfilo) [reply transformer_engine/common/util/rocm_cast_gated_kernels.cuh]: Truncated after_act_elt and after_dact_elt are used more than once interleaving with original after_dact_reg[stage] so if not use those variables will need to make truncation twice: for ROWWISE and for COLWISE

File: tests/jax/test_distributed_layernorm_mlp.py (hunk 7 above)
Comment (ipanfilo) [reply tests/jax/test_distributed_layernorm_mlp.py:238]: So I'll guard it with is_hip_extension() to avoid further confusion and distinct with upstream fix when it is done

File: tests/jax/test_distributed_softmax.py (hunk 8 above)
Comment (ipanfilo) [reply tests/jax/test_distributed_softmax.py]: It is not removed at line 179 but moved here but it indeed can be totally removed and the fill will reset to upstram

**Review by ipanfilo — COMMENTED** [review]
Dummy comment for dummy GitHub letting me submit my comments

File: transformer_engine/common/util/rocm_cast_gated_kernels.cuh (hunk 15 above)
Comment (wangye805) [reply transformer_engine/common/util/rocm_cast_gated_kernels.cuh]: But both NV kernel (https://github.com/ROCm/TransformerEngine/blob/a406914a9d17889fcc64d38e2cfed563a2b1a7b5/transformer_engine/common/util/cast_gated_kernels.cuh#L515C9-L515C26) and gtest reference (https://github.com/ROCm/TransformerEngine/blob/a406914a9d17889fcc64d38e2cfed563a2b1a7b5/tests/cpp/operator/test_cast_mxfp8_gated_swiglu.cu#L89) is taking downcasting, saving it to cache, and using the cached values afterwards. 

In other words,

File: transformer_engine/common/util/rocm_cast_gated_kernels.cuh
Comment (wangye805) [inline transformer_engine/common/util/rocm_cast_gated_kernels.cuh:351]: In gtest reference, the downcasted cache is used: https://github.com/ROCm/TransformerEngine/blob/a406914a9d17889fcc64d38e2cfed563a2b1a7b5/tests/cpp/operator/test_cast_mxfp8_gated_swiglu.cu#L167-L176

Same behavior in the NV kernel: https://github.com/ROCm/TransformerEngine/blob/a406914a9d17889fcc64d38e2cfed563a2b1a7b5/transformer_engine/common/util/cast_gated_kernels.cuh#L631-L633

**Review by wangye805 — CHANGES_REQUESTED** [review]

File: transformer_engine/pytorch/tensor/mxfp8_tensor.py (hunk 4 above)
Comment (wenchenvincent) [reply transformer_engine/pytorch/tensor/mxfp8_tensor.py:116]: Hmm, I think we should initialize the data tensor to 0. The reason is that we need to pad it to the multiple of 32 elements and padding position should be 0 while the non-padding position will be filled with meaningful data. The padding only happen to the data tensor I think and no padding is needed for the scale tensor.

Comment (ipanfilo) [reply transformer_engine/pytorch/tensor/mxfp8_tensor.py:116]: Look at the very next lines: scale tensor sizes are rounded up. It is done by NV tiling and alignment requirements. ROCm might not need it but many places relate on it so we follow the same scheme

Comment (wenchenvincent) [reply transformer_engine/pytorch/tensor/mxfp8_tensor.py:116]: Thanks. I see it is padded to 128 elements so it makes sense to also pad scale factors. But then my only remaining question is: shouldn't we also pad the data tensor?

**Review by wenchenvincent — APPROVED** [review]
LGTM.

**Review by wangye805 — APPROVED** [review]
LGTM
